src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -15,32 +16,67 @@ namespace System.Net.Http.Headers
     // Use HeaderDescriptor.TryGet to resolve an arbitrary header name to a HeaderDescriptor.
     internal readonly struct HeaderDescriptor : IEquatable<HeaderDescriptor>
     {
-        private readonly string _headerName;
-        private readonly KnownHeader? _knownHeader;
+        /// <summary>
+        /// Either a <see cref="KnownHeader"/> or <see cref="string"/>
+        /// </summary>
+        private readonly object _descriptor;
 
         public HeaderDescriptor(KnownHeader knownHeader)
         {
-            _knownHeader = knownHeader;
-            _headerName = knownHeader.Name;
+            _descriptor = knownHeader;
         }
 
         // This should not be used directly; use static TryGet below
         internal HeaderDescriptor(string headerName)
         {
-            _headerName = headerName;
-            _knownHeader = null;
+            _descriptor = headerName;
         }
 
-        public string Name => _headerName;
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public string Name
+        {
+            get
+            {
+                object? descriptor = _descriptor;
+                return descriptor is KnownHeader knownHeader ?
+                    knownHeader.Name :
+                    Unsafe.As<string>(descriptor);
```

Comment:
Why Unsafe.As?

Reply:
To avoid the cost of the type check

Reply:
If you want to optimize it further you change it to [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYGuGNmkATJqQYBpLhADuXABIGjMmgG8aDD012yMUXlwBzBgA5bAEGFwYAmAw1PRi4gF8aZOotXTIGIWxXd09YbCMILkkATwYIYAArGDAMBgB9J1wwPzEMaA006k8vJnJGUIEACgBKPI83Ht7PSpq6hmbW3nboBgBeRqW2jqgumc9iAHZF1WXVqAZeXAsrWwdCmQYAazv7RyeAfgmDz1ebd6PKAsIYwBggH6/BgAVS4uGwADNBABBXAAHmYdAAfMNtitdqN9p5Uj8dP1BmEYA0IO0xj8plC5rV6niLhstmcdp1IYcTqzdiwAOIxAAqpTEMDGG02GHFMAgCOGlgBDycUFGDG+0yhHlh8KRLFRaOV9w+UBx/OgoxBlPBPIOesRKPRmItnPxVqJHlSqSAA=) 😄 

judging by the fact you never expect `_descriptor` to be null and `KnownFolder` is sealed

Reply:
Oh this is lovely. There are some more similar uses in a few other loops this could help with.

Reply:
@EgorBo could the JIT do this optimization / trick for us?
IIf not null and the type is sealed as pre-condition.
The C# pattern matching is used quite often, the `GetType`-trick is less known, so a lot of usecases could profit from this.

Reply:
How much does avoiding the type check here contribute to the RPS improvement?

Reply:
@gfoidl JIT doesn't know that `_descriptor` holds only either `KnownFolder` or `string` and the value is never null - I don't think it's possible to do an analyze like that (aka interprocedural) and still someone might change the value of the field to something different or null via reflection.

Type check against `string` or `KnownFolder` should be super cheap as both are `sealed`

Reply:
I would imagine it's negligible for RPS compared to all the other costs.

Reply:
> I would imagine it's negligible for RPS compared to all the other costs.

Which argues for not using Unsafe here.

The concern here isn't so much correctness in this PR as it is maintenance.

Reply:
@EgorBo what I meant specifically is that the JIT could turn
```c#
return descriptor is KnownHeader knownHeader ?
    knownHeader.Name :
    // else path
```
into machine code similar to what is output from
```c#
return descriptor.GetType() == typeof(KnownHeader) ?
    Unsafe.As<KnownHeader>(descriptor).Name :
    // else path
```
Here JIT should see that KnownHeader is sealed, so no sub-typing.
Or is this with dynamic PGO already covered?

Reply:
@gfoidl JIT does exactly that already. the only thing why `GetType` + `Unsafe` are better because they remove `nullcheck` from that and JIT can't figure out that by its own that `descriptor` is never null.